static/src/javascripts/lib/cmp.spec.js

```diff
@@ -0,0 +1,215 @@
+
+                const myCallBack = jest.fn();
+
+                init();
+
+                onConsentNotification('advertisement', myCallBack);
+
+                _.triggerConsentNotification();
+
+                expect(myCallBack).toHaveBeenCalledTimes(2);
+                expect(myCallBack.mock.calls).toEqual([
+                    [true], // First call
+                    [true], // Second call
+                ]);
+            });
+        });
+
+        describe('if cmpIsReady is FALSE when onConsentNotification called waits to', () => {
+            it('execute functional callback', () => {
+                const myCallBack = jest.fn();
+
+                onConsentNotification('functional', myCallBack);
+
+                expect(myCallBack).not.toBeCalled();
+
+                init();
+
+                expect(myCallBack).toHaveBeenCalledTimes(1);
+                expect(myCallBack).toBeCalledWith(true);
+            });
+
+            it('execute performance callback', () => {
+                const myCallBack = jest.fn();
+
+                onConsentNotification('performance', myCallBack);
+
+                expect(myCallBack).not.toBeCalled();
+
+                init();
+
+                expect(myCallBack).toHaveBeenCalledTimes(1);
+                expect(myCallBack).toBeCalledWith(true);
+            });
+
+            it('execute advertisement callback with true if getAdConsentState true', () => {
+                getAdConsentState.mockReturnValue(true);
+
+                const myCallBack = jest.fn();
+
+                onConsentNotification('advertisement', myCallBack);
+
+                expect(myCallBack).not.toBeCalled();
+
+                init();
+
+                expect(myCallBack).toHaveBeenCalledTimes(1);
+                expect(myCallBack).toBeCalledWith(true);
+            });
+
+            it('execute advertisement callback with true if getAdConsentState false', () => {
+                getAdConsentState.mockReturnValue(false);
+
+                const myCallBack = jest.fn();
+
+                onConsentNotification('advertisement', myCallBack);
+
+                expect(myCallBack).not.toBeCalled();
+
+                init();
+
+                expect(myCallBack).toHaveBeenCalledTimes(1);
+                expect(myCallBack).toBeCalledWith(false);
+            });
+
+            it('execute advertisement callback each time consent nofication triggered', () => {
+                getAdConsentState.mockReturnValue(true);
+
+                const myCallBack = jest.fn();
+
+                onConsentNotification('advertisement', myCallBack);
+
+                expect(myCallBack).not.toBeCalled();
+
+                init();
+
+                _.triggerConsentNotification();
+
+                expect(myCallBack).toHaveBeenCalledTimes(2);
+                expect(myCallBack.mock.calls).toEqual([
+                    [true], // First call
+                    [true], // Second call
+                ]);
+            });
```

Comment:
I don't see any tests to check that it can be re-triggered after a change to the user's consent. We'd want to make sure that the callbacks are invoked with the new value.

Reply:
The test `execute advertisement callback each time consent nofication triggered` does via the call `_.triggerConsentNotification();`, calling this simulates what should be executed after a change to the user's consent has been made. 

Currently this module only exposes `triggerConsentNotification` as  part of `_` for testing purposes. However I foresee `triggerConsentNotification` being on the public interface for the module once we implement the CMP UI, as the UI will need to call  `triggerConsentNotification` whenever the user submit's their consent via the UI and it has been saved in the cookie. 

We could just expose `triggerConsentNotification` now as part of the actual interface (not just for testing purposes) as we're going to need this exposed soon anyway. What do you think?

Reply:
I mean testing that the consent starts at `X`, then is changed to `Y`. The callbacks should be invoked with `X` and then `Y`. All the examples I spotted checked repeated invocations, but didn't test that the callback arguments can change between invocations.

Sorry for not being clear!

Reply:
@GHaberis triggerConsentNotification should only need to be exposed as a test function. Our saveCookie function will trigger it internally every time it's called.

@adamnfish As we don't yet have a way to change the consent state the current tests are the best we can do. But we'll add the tests you described when functions such as saveCookie and readCookie are implemented on the library.

Reply:
> As we don't yet have a way to change the consent state the current tests are the best we can do. But we'll add the tests you described when functions such as saveCookie and readCookie are implemented on the library.

This is correct, whilst there's no UI there's nothing to update the consent state right now.

> triggerConsentNotification should only need to be exposed as a test function. Our saveCookie function will trigger it internally every time it's called.

That depends on whether `saveCookie` and `triggerConsentNotification` are in the same module, I'd imagined they won't be as the scope of such a module would be quite broad, and prefer them to be in separate modules  with distinct purposes, in which case `triggerConsentNotification` would need to be exposed.

Reply:
I agree. But I didn't mean the function that saves the actual cookie. Just the externally facing function (whatever its name is) that causes a change in consent and subsequently triggers the cookie save. That one will most likely be in the same module as triggerConsentNotification and that's where the cookie state will be kept.

Reply:
> we don't yet have a way to change the consent state

There is an existing mechanism for the advertising purpose, but no worries if you want to introduce all of this together later.

> But we'll add the tests you described when functions such as saveCookie and readCookie are implemented on the library

Sounds good.